wremnants/muon_efficiencies.py

```diff
@@ -66,52 +66,90 @@ def make_muon_efficiency_helpers(filename = data_dir + "/testMuonSF/scaleFactorP
+        sf_tracking.view(flow=True)[:, :, axis_charge.index(-1.), axis_nom_alt.index(nom_syst)] = hist_hist.view(flow=True)[...]
+        sf_tracking.view(flow=True)[:, :, axis_charge.index(1.), axis_nom_alt.index(nom_syst)] = hist_hist.view(flow=True)[...]
+
+        ## FIXME: avoid code duplication?
+        ## now again for reco
+        hist_name = f"fullSF2D_{nom_syst_tag}_reco_{eratag}"
+        hist_root = fin.Get(hist_name)
+        hist_hist = narf.root_to_hist(hist_root, axis_names = ["SF eta", "SF pt"])
 
+        if sf_reco is None:
+            axis_eta_eff_reco = hist_hist.axes[0]
+            axis_pt_eff_reco = hist_hist.axes[1]
+            sf_reco = hist.Hist(axis_eta_eff_reco, axis_pt_eff_reco, axis_charge, axis_nom_alt,  name = "sf_reco", storage = hist.storage.Weight())
 
         # same SFs for + and - for now
-        sf_tracking_reco.view(flow=True)[:, axis_charge.index(-1.), axis_nom_alt.index(nom_syst)] = hist_hist.view(flow=True)[...]
-        sf_tracking_reco.view(flow=True)[:, axis_charge.index(1.), axis_nom_alt.index(nom_syst)] = hist_hist.view(flow=True)[...]
+        sf_reco.view(flow=True)[:, :, axis_charge.index(-1.), axis_nom_alt.index(nom_syst)] = hist_hist.view(flow=True)[...]
+        sf_reco.view(flow=True)[:, :, axis_charge.index(1.), axis_nom_alt.index(nom_syst)] = hist_hist.view(flow=True)[...]
 
+        
     # set overflow and underflow equal to adjacent bins
-    sf_tracking_reco.view(flow=True)[0, ...] = sf_tracking_reco.view(flow=True)[1, ...]
-    sf_tracking_reco.view(flow=True)[axis_eta_eff_tracking_reco.extent-1, ...] = sf_tracking_reco.view(flow=True)[axis_eta_eff_tracking_reco.extent-2, ...]
+    sf_tracking.view(flow=True)[0, ...] = sf_tracking.view(flow=True)[1, ...]
+    sf_tracking.view(flow=True)[axis_eta_eff_tracking.extent-1, ...] = sf_tracking.view(flow=True)[axis_eta_eff_tracking.extent-2, ...]
+    sf_tracking.view(flow=True)[:, 0, ...] = sf_tracking.view(flow=True)[:, 1, ...]
+    sf_tracking.view(flow=True)[:, axis_pt_eff_tracking.extent-1, ...] = sf_tracking.view(flow=True)[:, axis_pt_eff_tracking.extent-2, ...]
+    # set overflow and underflow equal to adjacent bins
+    sf_reco.view(flow=True)[0, ...] = sf_reco.view(flow=True)[1, ...]
+    sf_reco.view(flow=True)[axis_eta_eff_reco.extent-1, ...] = sf_reco.view(flow=True)[axis_eta_eff_reco.extent-2, ...]
+    sf_reco.view(flow=True)[:, 0, ...] = sf_reco.view(flow=True)[:, 1, ...]
+    sf_reco.view(flow=True)[:, axis_pt_eff_reco.extent-1, ...] = sf_reco.view(flow=True)[:, axis_pt_eff_reco.extent-2, ...]
 
+    
     fin.Close()
 
     # TODO implement a convenient way to use enums for the bool/category/nom-var axes such that
     # the C++ code is less error-prone?
 
-    netabins = axis_eta_eff.size
+    netabins = axis_eta_eff.size  # this is the same also for reco and tracking, should one check explicitly?
 
     # exclude pt bins outside of analysis range
     nptbins = 0
+    nptbins_tracking = 0
```

Comment:
Loops in bumpy are evil ;)

nptbins = np.count_nonzero(axis_pt_eff.edges[1:] > max_pt) 

Or you might not need the offset if you want to check the first bin boundary not the end edge

Reply:
ok, for now it was just removing the upper part, bins below the actual minus pt in the analysis are currently kept, but we may want to improve that too, especially if we raise the lower pt cut (currently we only have one dummy nuisance for pt in 24-26)

Reply:
I would quickly test the right use of np.count_nonzero, maybe adding or subtracting 1, to get the desired result. But it's much more compact and much faster than a python loop

Reply:
Done